Disha Jain - dj9am
Sameer Gupta - sg4vh
Jason Quinn - jtq5ba
Sindhu Ranga - sgr7va

## Project 2: Kaggle Zillow prize challenge replication

<div class="alert alert-block alert-danger">
<b>Due: 9:29am, Tuesday, 4 February 2020</b>
</div>

In your second project you are tasked with building a model to improve the Zestimate residual error using the data for 2017 in the Kaggle Zillow prize competition. The following description is adapted from the Kaggle. For more details, reference: https://www.kaggle.com/c/zillow-prize-1/

### Data

You can only use the following data.

<div class="alert alert-block alert-warning">
Download the data you need for this assignment from:
Collab/Resources/Datasets
</div>

This file contains two `.csv` (comma-separated values) files, and one Excel directory file. Unzip the file to extract the CSV files into a directory of your choice.

### Data description

(Train/Test split)

- You are provided with a full list of real estate properties in three counties (Los Angeles, Orange and Ventura, California) data in 2017 in the file `properties_2017.csv`.

- Not all the properties are sold in each time period. If a property was not sold in the time period, it will not have a row in `train_2017.csv` and so will not be used in predictions.

(File descriptions)

- properties_2017.csv - all the properties with their home features for 2017 (released on 10/2/2017)
- train_2017.csv - the training set with transactions from 1/1/2017 to 9/15/2017 (released on 10/2/2017)

(Data fields)

- Please refer to zillow_data_dictionary.xlsx

### Instruction Overview

Zillow Prize is challenging the data science community to help push the accuracy of the Zestimate even further. In the competition, Zillow is asking you to predict the log-error between their Zestimate and the actual sale price, given all the features of a home. 

The log error is defined as

    logerror=log(Zestimate)−log(SalePrice)

and it is recorded in the transactions file `train_2017.csv`. Using this training set and features of the home, set up your model for log error prediction. And then for each property (unique parcelid) in the `properties_2017.csv` dataset, you must predict a log error for the next period. Your program should write to output.csv in the following format.
(Example)

|   parcelid    |    logerror   |
| ------------- | ------------- |
|   10754147    |    0.1234     |
|   10759547    |   -0.3212     |
|       ...     |       ...     |

Your algorithm also needs to output the mean of all logerrors. 
Your answers should be in the form of a clear argument that includes both well-written prose, code and the numerical results (when the notebook is run). 


### Assignment

You should complete the assignment by inserting cells in the notebook with your answers to these questions, including both prose and code you used for your analysis.

<div class="alert alert-block alert-warning">
 Construct a model and predict the log-error for each property (unique parcelid) given all the features of a home.
</div>

Let's start with the data loading.

In [1]:
# Import the libraries and give them abbreviated names:
import pandas as pd
import numpy as np
import statsmodels.api as sm

# load the data, use the directory where you saved the data:
df_properties = pd.read_csv('properties_2017.csv') 
df_train = pd.read_csv('train_2017.csv', parse_dates=["transactiondate"])

/usr/local/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3058: DtypeWarning: Columns (49) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [2]:
df_properties.head()

parcelid  airconditioningtypeid  architecturalstyletypeid  basementsqft  \
0  10754147                    NaN                       NaN           NaN   
1  10759547                    NaN                       NaN           NaN   
2  10843547                    NaN                       NaN           NaN   
3  10859147                    NaN                       NaN           NaN   
4  10879947                    NaN                       NaN           NaN   

   bathroomcnt  bedroomcnt  buildingclasstypeid  buildingqualitytypeid  \
0          0.0         0.0                  NaN                    NaN   
1          0.0         0.0                  NaN                    NaN   
2          0.0         0.0                  5.0                    NaN   
3          0.0         0.0                  3.0                    6.0   
4          0.0         0.0                  4.0                    NaN   

   calculatedbathnbr  decktypeid  ...  numberofstories  fireplaceflag  \
0                NaN         NaN  ...              NaN            NaN   
1                NaN         NaN  ...              NaN            NaN   
2                NaN         NaN  ...              1.0            NaN   
3                NaN         NaN  ...              1.0            NaN   
4                NaN         NaN  ...              1.0            NaN   

   structuretaxvaluedollarcnt  taxvaluedollarcnt  assessmentyear  \
0                         NaN                9.0          2016.0   
1                         NaN            27516.0          2015.0   
2                    660680.0          1434941.0          2016.0   
3                    580059.0          1174475.0          2016.0   
4                    196751.0           440101.0          2016.0   

   landtaxvaluedollarcnt  taxamount  taxdelinquencyflag  taxdelinquencyyear  \
0                    9.0        NaN                 NaN                 NaN   
1                27516.0        NaN                 NaN                 NaN   
2               774261.0   20800.37                 NaN                 NaN   
3               594416.0   14557.57                 NaN                 NaN   
4               243350.0    5725.17                 NaN                 NaN   

   censustractandblock  
0                  NaN  
1                  NaN  
2                  NaN  
3                  NaN  
4                  NaN  

[5 rows x 58 columns]

In [3]:
df_train.head()

parcelid  logerror transactiondate
0  14297519  0.025595      2017-01-01
1  17052889  0.055619      2017-01-01
2  14186244  0.005383      2017-01-01
3  12177905 -0.103410      2017-01-01
4  10887214  0.006940      2017-01-01

### Join the two dataframes on parcel_id

In [4]:
prop_sorted = df_properties.sort_values(by='parcelid', axis=0)
train_sorted = df_train.sort_values(by='parcelid',axis=0)

In [5]:
merged_inner = pd.merge(left=df_train,right=df_properties, left_on='parcelid', right_on='parcelid')
merged_inner.head()


parcelid  logerror transactiondate  airconditioningtypeid  \
0  14297519  0.025595      2017-01-01                    NaN   
1  17052889  0.055619      2017-01-01                    NaN   
2  14186244  0.005383      2017-01-01                    NaN   
3  12177905 -0.103410      2017-01-01                    NaN   
4  10887214  0.006940      2017-01-01                    1.0   

   architecturalstyletypeid  basementsqft  bathroomcnt  bedroomcnt  \
0                       NaN           NaN          3.5         4.0   
1                       NaN           NaN          1.0         2.0   
2                       NaN           NaN          2.0         3.0   
3                       NaN           NaN          3.0         4.0   
4                       NaN           NaN          3.0         3.0   

   buildingclasstypeid  buildingqualitytypeid  ...  numberofstories  \
0                  NaN                    NaN  ...              NaN   
1                  NaN                    NaN  ...              1.0   
2                  NaN                    NaN  ...              1.0   
3                  NaN                    8.0  ...              NaN   
4                  NaN                    8.0  ...              NaN   

   fireplaceflag  structuretaxvaluedollarcnt  taxvaluedollarcnt  \
0            NaN                    485713.0          1023282.0   
1            NaN                     88000.0           464000.0   
2            NaN                     85289.0           564778.0   
3            NaN                    108918.0           145143.0   
4            NaN                     73681.0           119407.0   

   assessmentyear  landtaxvaluedollarcnt  taxamount  taxdelinquencyflag  \
0          2016.0               537569.0   11013.72                 NaN   
1          2016.0               376000.0    5672.48                 NaN   
2          2016.0               479489.0    6488.30                 NaN   
3          2016.0                36225.0    1777.51                 NaN   
4          2016.0                45726.0    1533.89                 NaN   

   taxdelinquencyyear  censustractandblock  
0                 NaN         6.059063e+13  
1                 NaN         6.111001e+13  
2                 NaN         6.059022e+13  
3                 NaN         6.037300e+13  
4                 NaN         6.037124e+13  

[5 rows x 60 columns]

In [6]:
train_df = merged_inner
train_df.dropna(subset=['parcelid'])
y_train=train_df[['logerror']]
X_train=train_df.drop(columns=['logerror','parcelid'], axis=0)

In [7]:
corr_matrix = merged_inner.corr()
corr_matrix["logerror"].sort_values(ascending=False)

logerror                        1.000000
basementsqft                    0.372067
buildingclasstypeid             0.315372
finishedsquarefeet6             0.072870
finishedsquarefeet12            0.045921
calculatedfinishedsquarefeet    0.040516
garagetotalsqft                 0.035015
bedroomcnt                      0.031638
calculatedbathnbr               0.029330
garagecarcnt                    0.029002
fullbathcnt                     0.027133
bathroomcnt                     0.025817
fireplacecnt                    0.023242
poolsizesum                     0.021174
longitude                       0.015876
threequarterbathnbr             0.015540
parcelid                        0.015407
roomcnt                         0.014567
lotsizesquarefeet               0.011012
airconditioningtypeid           0.009341
structuretaxvaluedollarcnt      0.008433
numberofstories                 0.008204
fips                            0.006413
rawcensustractandblock          0.006333
yearbuilt       

In [8]:
numerical_df = merged_inner[['logerror','finishedsquarefeet13','finishedsquarefeet13','finishedfloor1squarefeet','landtaxvaluedollarcnt','unitcnt','taxamount','taxvaluedollarcnt','numberofstories','lotsizesquarefeet','bathroomcnt','fullbathcnt','fireplacecnt','garagecarcnt','bedroomcnt','garagetotalsqft','basementsqft']]

In [9]:
merged_inner['finishedsquarefeet13_log'] = np.log10(merged_inner['finishedsquarefeet13'])
merged_inner['finishedfloor1squarefeet_log'] = np.log10(merged_inner['finishedfloor1squarefeet'])
merged_inner['landtaxvaluedollarcnt_log'] = np.log10(merged_inner['landtaxvaluedollarcnt'])
merged_inner['taxvaluedollarcnt_log'] = np.log10(merged_inner['taxvaluedollarcnt'])

# numerical_df = merged_inner[['logerror','finishedsquarefeet13','finishedfloor1squarefeet','landtaxvaluedollarcnt','unitcnt','taxamount','taxvaluedollarcnt','numberofstories','lotsizesquarefeet','bathroomcnt','fullbathcnt','fireplacecnt','garagecarcnt','bedroomcnt','garagetotalsqft','basementsqft']]
corr_matrix = merged_inner.corr()
corr_matrix["logerror"].sort_values(ascending=False)

logerror                        1.000000
basementsqft                    0.372067
buildingclasstypeid             0.315372
finishedsquarefeet6             0.072870
finishedsquarefeet12            0.045921
calculatedfinishedsquarefeet    0.040516
garagetotalsqft                 0.035015
bedroomcnt                      0.031638
calculatedbathnbr               0.029330
garagecarcnt                    0.029002
fullbathcnt                     0.027133
bathroomcnt                     0.025817
fireplacecnt                    0.023242
poolsizesum                     0.021174
longitude                       0.015876
threequarterbathnbr             0.015540
parcelid                        0.015407
roomcnt                         0.014567
lotsizesquarefeet               0.011012
airconditioningtypeid           0.009341
structuretaxvaluedollarcnt      0.008433
numberofstories                 0.008204
fips                            0.006413
rawcensustractandblock          0.006333
yearbuilt       

In [10]:
X_train.head()

transactiondate  airconditioningtypeid  architecturalstyletypeid  \
0      2017-01-01                    NaN                       NaN   
1      2017-01-01                    NaN                       NaN   
2      2017-01-01                    NaN                       NaN   
3      2017-01-01                    NaN                       NaN   
4      2017-01-01                    1.0                       NaN   

   basementsqft  bathroomcnt  bedroomcnt  buildingclasstypeid  \
0           NaN          3.5         4.0                  NaN   
1           NaN          1.0         2.0                  NaN   
2           NaN          2.0         3.0                  NaN   
3           NaN          3.0         4.0                  NaN   
4           NaN          3.0         3.0                  NaN   

   buildingqualitytypeid  calculatedbathnbr  decktypeid  ...  numberofstories  \
0                    NaN                3.5         NaN  ...              NaN   
1                    NaN                1.0         NaN  ...              1.0   
2                    NaN                2.0         NaN  ...              1.0   
3                    8.0                3.0         NaN  ...              NaN   
4                    8.0                3.0         NaN  ...              NaN   

   fireplaceflag  structuretaxvaluedollarcnt  taxvaluedollarcnt  \
0            NaN                    485713.0          1023282.0   
1            NaN                     88000.0           464000.0   
2            NaN                     85289.0           564778.0   
3            NaN                    108918.0           145143.0   
4            NaN                     73681.0           119407.0   

   assessmentyear  landtaxvaluedollarcnt  taxamount  taxdelinquencyflag  \
0          2016.0               537569.0   11013.72                 NaN   
1          2016.0               376000.0    5672.48                 NaN   
2          2016.0               479489.0    6488.30                 NaN   
3          2016.0                36225.0    1777.51                 NaN   
4          2016.0                45726.0    1533.89                 NaN   

   taxdelinquencyyear  censustractandblock  
0                 NaN         6.059063e+13  
1                 NaN         6.111001e+13  
2                 NaN         6.059022e+13  
3                 NaN         6.037300e+13  
4                 NaN         6.037124e+13  

[5 rows x 58 columns]

In [11]:
from sklearn.model_selection import train_test_split

train_df = merged_inner
train_df.dropna(subset=['parcelid'])
y_train=train_df[['logerror']]
X_train=train_df.drop(columns=['logerror','parcelid'], axis=0)

variable = 'calculatedbathnbr'
variable1 = 'garagetotalsqft'
variable2 = 'finishedsquarefeet6'
X_train.loc[X_train[variable].isnull(), variable] =0
X_train.loc[X_train[variable1].isnull(), variable1] =X_train[variable1].mean()
X_train.loc[X_train[variable2].isnull(), variable2] = X_train[variable2].mean()
my_variable = X_train[[variable, variable1, variable2]]

X, X_test, y, y_test = train_test_split(my_variable, y_train, test_size=0.16, random_state=42)


y_test.head()

logerror
59756  0.052298
69100  0.005054
54038  0.011440
75563 -0.069159
36562  0.028567

In [12]:
# basement_median = np.median(np.asarray(X_train['basementsqft']))
# my_variable = X_train['basementsqft'].fillna(basement_median)


# my_variable = np.asarray(X_train[variable]).reshape(-1,1)
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

lin_reg = LinearRegression()

lin_reg.fit(X, y[['logerror']])

predictions = lin_reg.predict(X_test)
print("SECOND")
print(predictions)

my_sum = np.sum(np.subtract(predictions,y_test[['logerror']])**2)

print("SSE: "+str(my_sum))
    
mse = mean_squared_error(y_test[['logerror']], predictions)
rmse = np.sqrt(mse)
print("MSE: " + str(mse))
print("RMSE: " + str(rmse))

from sklearn.model_selection import cross_val_score

scores = cross_val_score(lin_reg, X_test, y_test[['logerror']],
                         scoring="neg_mean_squared_error", cv=10)
rmse_scores = np.sqrt(-scores)
print("Scores:", rmse_scores)
print(pd.Series(rmse_scores).describe())


SECOND
[[0.01557534]
 [0.02221417]
 [0.01125216]
 ...
 [0.02414781]
 [0.01557534]
 [0.01125216]]
SSE: logerror    411.19125
dtype: float64
MSE: 0.03310985184221622
RMSE: 0.181961127283319
Scores: [0.1968172  0.19632083 0.24978023 0.14404809 0.14006549 0.16915137
 0.15284596 0.16359686 0.21995397 0.15526237]
count    10.000000
mean      0.178784
std       0.035937
min       0.140065
25%       0.153450
50%       0.166374
75%       0.196693
max       0.249780
dtype: float64


In [13]:
print(lin_reg.coef_)

[[4.32318188e-03 1.91450261e-05 2.59780397e-05]]


In [14]:
from sklearn.preprocessing import PolynomialFeatures

train_df = merged_inner
train_df.dropna(subset=['parcelid'])
y_train=train_df[['logerror']]
X_train=train_df.drop(columns=['logerror','parcelid'], axis=0)

variable = 'calculatedbathnbr'
variable1 = 'garagetotalsqft'
variable2 = 'finishedsquarefeet6'
X_train.loc[X_train[variable].isnull(), variable] =0
X_train.loc[X_train[variable1].isnull(), variable1] =X_train[variable1].mean()
X_train.loc[X_train[variable2].isnull(), variable2] = X_train[variable2].mean()
my_variable = X_train[[variable, variable1, variable2]]

poly = PolynomialFeatures(degree=5)
my_variable = poly.fit_transform(my_variable)

X, X_test, y, y_test = train_test_split(my_variable, y_train, test_size=0.16, random_state=42)


y_test.head()

logerror
59756  0.052298
69100  0.005054
54038  0.011440
75563 -0.069159
36562  0.028567

In [15]:
lin_reg = LinearRegression()

lin_reg.fit(X, y[['logerror']])

predictions = lin_reg.predict(X_test)
print("SECOND")
print(predictions)

my_sum = np.sum(np.subtract(predictions,y_test[['logerror']])**2)

print("SSE: "+str(my_sum))
    
mse = mean_squared_error(y_test[['logerror']], predictions)
rmse = np.sqrt(mse)
print("MSE: " + str(mse))
print("RMSE: " + str(rmse))

from sklearn.model_selection import cross_val_score

scores = cross_val_score(lin_reg, X_test, y_test[['logerror']],
                         scoring="neg_mean_squared_error", cv=10)
rmse_scores = np.sqrt(-scores)
print("Scores:", rmse_scores)
print(pd.Series(rmse_scores).describe())


SECOND
[[0.01397382]
 [0.01984625]
 [0.0158147 ]
 ...
 [0.0235788 ]
 [0.01397382]
 [0.0158147 ]]
SSE: logerror    411.477141
dtype: float64
MSE: 0.033132872271195386
RMSE: 0.18202437273946417
Scores: [0.19775117 0.19639014 0.72192496 0.14474301 0.23196361 0.1702943
 2.8190572  0.16375651 0.2203764  0.15556612]
count    10.000000
mean      0.502182
std       0.831826
min       0.144743
25%       0.165391
50%       0.197071
75%       0.229067
max       2.819057
dtype: float64


In [32]:
def get_predictions(X_train, y_train, X_test, y_test):
    lin_reg = LinearRegression()

    lin_reg.fit(X_train, y_train[['logerror']])

    predictions = lin_reg.predict(X_test)
    print(y_test[['logerror']].shape)
    print(X_test.shape)
    mse = mean_squared_error(y_test[['logerror']], predictions)
    rmse = np.sqrt(mse)
    print("MSE: " + str(mse))
    print("RMSE: " + str(rmse))
    i = 0
    for logerror in y_test['logerror']:
        logerror = predictions[i]
        i+=1
    return y_test

In [33]:
import datetime
train_df = merged_inner
train_df.dropna(subset=['parcelid'])

variable = 'calculatedbathnbr'
variable1 = 'garagetotalsqft'
variable2 = 'finishedsquarefeet6'

train_df.loc[train_df[variable].isnull(), variable] =0
train_df.loc[train_df[variable1].isnull(), variable1] =train_df[variable1].mean()
train_df.loc[train_df[variable2].isnull(), variable2] = train_df[variable2].mean()


split_date = pd.Timestamp(datetime.date(2017,8,5))

df = train_df.loc[train_df['transactiondate'] <= split_date]
test_df = train_df.loc[train_df['transactiondate'] > split_date]

print(test_df.shape)
y=df[['logerror']]
X=df[[variable, variable1, variable2]]
y_test=test_df[['logerror']]
X_test=test_df[[variable, variable1, variable2]]

get_predictions(X, y, X_test, y_test)

(13708, 64)
(13708, 1)
(13708, 3)
MSE: 0.04266417172816405
RMSE: 0.20655307242489532


logerror
2681  -0.163064
3905  -0.064447
5064  -0.027589
6062   0.127325
6132  -0.792479
...         ...
77608 -0.002245
77609  0.020615
77610  0.013209
77611  0.037129
77612  0.007204

[13708 rows x 1 columns]

(1) Provide an argument for which variables in `properties_2017.csv` can potentially be good predictors for the value of interest (log-error), and explain why each will be useful. Explanations of why certain variables could be particularly poor predictors are also welcomed (but not required).


<div class="alert alert-block alert-info">We started by printing out all the log error predictions that had parcelid variables associated with them, and then running a correlation matrix between logerror and all the variables in that dataset. We found that 'calculatedbathnbr', 'garagetotalsqft', and 'finishedsquarefeet6' were the most correlated, thus we decided to use them in our model. We then ran a linear regression using these three variables to predict logerror, but we also decided to do a quartic polynomial regression to double check our models functional form, and see if other models/estimates were better suited for the data. We thought that any of the predictors with betas/correlations coefficients close to zero or negative would be particularly poor to predict.</div>

(2) You can see that some properties such as parcelid have more detailed home information available, such as number of bedrooms or bathrooms, and other areas do not in `properties_2017.csv`. Explain how you can handle this missing information when you construct a model.


<div class="alert alert-block alert-info">When there is missing information, there are a couple of different ways to handle it, depending on the type of variable, we could do various transformations of categorical data into dummy variables, or even filling in missing numerical values with mean, median, or mode depending on what made the most sense for that variable. For some numerical values like basement sq ft with no value we put in zero, because a Null value in that column may have implied there was no basement in that house. For other numerical values like finished square feet, we used the mean of that variable, as it was highly unlikely a house had no finished square feet.
</div>

(3) Construct your model and report the predicted log-error for each property for the next period. This should be a python script `project2.py` that reads from `properties_2017.csv` and writes to `output.csv` that lists parcelID and log-error predicted value for each parcel. 

Evaluate your model's ability to predict log-error values by comparing the predicted log-error values for the next period and the actual log-error testing values. 

Discuss why the model can or cannot predict the log-error values for the test data. 


<div class="alert alert-block alert-info">We split the data into two sections before August 2017 and after August 2017, to see how accurately we could predict the loc errors for the period after August 2017. After regressing, we noticed that the model was good at predicting the log error for the next period. Our MSE for the next period prediction was .042, and our MSE for the training data was similar, clocking in at .033. Since this is linear, it is hard to imagine that overfitting is possible, and based on our results, it is clear our model is not overfitted. The RMSE for the next period was .207, which means that on average, our log error predictions were off by around .207 in one direction. This is a pretty good residual, since it is quite small. Therefore, we believe that our model predicts log-error pretty well.</div>